src/Build/Evaluation/ConditionEvaluator.cs

```diff
@@ -194,41 +225,45 @@ internal static void UpdateConditionedPropertiesTable
 
-            Hashtable cachedExpressionTreesForCurrentOptions = s_cachedExpressionTrees[(int)options];
-
-            // We only need to lock on writes to the table
-            if (cachedExpressionTreesForCurrentOptions == null)
-            {
-                // Given property functions, casing in conditional expressions isn't necessarily ignored.
-                cachedExpressionTreesForCurrentOptions = new Hashtable(50, StringComparer.Ordinal);
-
-                lock (s_cachedExpressionTrees)
-                {
-                    s_cachedExpressionTrees[(int)options] = cachedExpressionTreesForCurrentOptions;
-                }
-            }
+            ExpressionTreeForCurrentOptionsWithSize cachedExpressionTreesForCurrentOptions = s_cachedExpressionTrees.GetOrAdd(
+                (int)options,
+                _ => new ExpressionTreeForCurrentOptionsWithSize(new ConcurrentDictionary<string, ConcurrentStack<GenericExpressionNode>>(StringComparer.Ordinal)));
 
-            // VS stress tests could fill up this cache without end, for example if they use
-            // random configuration names - those appear in conditional expressions.
-            // So if we hit a limit that we should never hit in normal circumstances in VS,
-            // and rarely, periodically hit in normal circumstances in large tree builds,
-            // just clear out the cache. It can start repopulating again. Some kind of prioritized
-            // aging isn't worth it: although the hit rate of these caches is excellent (nearly 100%)
-            // the cost of reparsing expressions should the cache be cleared is not particularly large.
-            // Loading Australian Government in VS, there are 3 of these tables, two with about 50
-            // entries and one with about 650 entries. So 3000 seems large enough.
-            if (cachedExpressionTreesForCurrentOptions.Count > 3000) // threadsafe
+            if (cachedExpressionTreesForCurrentOptions.OptimisticSize > 3000)
             {
-                lock (cachedExpressionTreesForCurrentOptions)
-                {
-                    cachedExpressionTreesForCurrentOptions.Clear();
-                }
+                // VS stress tests could fill up this cache without end, for example if they use
+                // random configuration names - those appear in conditional expressions.
+                // So if we hit a limit that we should never hit in normal circumstances in VS,
+                // and rarely, periodically hit in normal circumstances in large tree builds,
+                // just clear out the cache. It can start repopulating again. Some kind of prioritized
+                // aging isn't worth it: although the hit rate of these caches is excellent (nearly 100%)
+                // the cost of reparsing expressions should the cache be cleared is not particularly large.
+                // Loading Australian Government in VS, there are 3 of these tables, two with about 50
```

Comment:
We should consider checking this out on Roslyn or another modern large solution. Things might have grown.